test/framework/finalizers_helpers.go

```diff
@@ -0,0 +1,170 @@
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	clusterctlcluster "sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1beta1"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/patch"
+)
+
+// finalizerAssertion contains list of expected Finalizers corresponding to a resource Kind.
+var finalizerAssertion = map[string][]string{
+	"Cluster":             {clusterv1.ClusterFinalizer},
+	"Machine":             {clusterv1.MachineFinalizer},
+	"MachineSet":          {clusterv1.MachineSetTopologyFinalizer},
+	"MachineDeployment":   {clusterv1.MachineDeploymentTopologyFinalizer},
+	"ClusterResourceSet":  {addonsv1.ClusterResourceSetFinalizer},
+	"DockerMachine":       {infrav1.MachineFinalizer},
+	"DockerCluster":       {infrav1.ClusterFinalizer},
+	"KubeadmControlPlane": {controlplanev1.KubeadmControlPlaneFinalizer},
+}
+
+// ValidateFinalizersResilience checks that expected finalizers are in place, deletes them, and verifies that expected finalizers are properly added again.
+func ValidateFinalizersResilience(ctx context.Context, proxy ClusterProxy, namespace, clusterName string) {
+	clusterKey := client.ObjectKey{Namespace: namespace, Name: clusterName}
+
+	// collect all objects where finalizers were initially set
+	objectsWithFinalizers := getObjectsWithFinalizers(ctx, proxy, namespace)
+
+	// Removes all the Finalizers.
+	// We are testing the worst-case scenario, i.e. all finalizers are deleted, even when reconciliation is paused.
+	// Deletion of the Finalizers can either happen from a system generated attack or by mistake, which is an edge case.
+	// The reconciliation loop should be able to recover from this, by adding the required finalizers back.
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, true)
+
+	// Once all Clusters are paused remove the Finalizers from all objects in the graph.
+	removeFinalizers(ctx, proxy, namespace)
+
+	// Unpause the cluster.
+	setClusterPause(ctx, proxy.GetClient(), clusterKey, false)
+
+	// Annotate the clusterClass, if one is in use, to speed up reconciliation. This ensures ClusterClass Finalizers
+	// are re-reconciled before asserting the owner reference graph.
+	forceClusterClassReconcile(ctx, proxy.GetClient(), clusterKey)
+
+	// The MachineDeployment Topology reconciler does not trigger reconciliations, when owner Cluster is in paused state.
+	// So, a manual update is required to trigger a reconciliation.
+	forceMachineDeploymentTopologyReconcile(ctx, proxy.GetClient(), clusterKey)
```

Comment:
When the owner `Cluster` object is paused, the Topology based `MachineDeployment` controller doesn't reconcile. 
The owner references tests were running since the other  `MachineDeployment`  controller sets these owner references: https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machinedeployment/machinedeployment_controller.go#L209

And watches for `Cluster` object changes as well https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machinedeployment/machinedeployment_controller.go#L92

Neither of these cases hold true in the `MachineDeplyoment` Topology based controller, but on the other hand the topology finalisers are only set in this controller - which IMO is a bug 🐛

Reply:
Everything should be unpaused at this point in the function AFAICT

Reply:
TLDR: `Cluster` updates do not trigger **topology** owned `MachineDeployment` reconciliation, since the responsible MD controller doesn't watch for `Cluster` resources.

Reply:
I have created a follow-up feature request here https://github.com/kubernetes-sigs/cluster-api/issues/9532